### ISSUE TITLE
useEffect is broken for React Native with JSC

### ISSUE THREAD

**Issue by bvaughn** [facebook/react#14352]
As reported in open source (facebook/react-native/issues/21967), the `useEffect` hook is broken for React Native when using JavaScriptCore (which affects both iOS and Android).

This is because the [`setTimeout` branch of `scheduler`](https://github.com/facebook/react/blob/c2a2d8a539bf02e40c43d36adc2826e228f30955/packages/scheduler/src/Scheduler.js#L463-L495) specifies a [5000ms delay](https://github.com/facebook/react/blob/c2a2d8a539bf02e40c43d36adc2826e228f30955/packages/scheduler/src/Scheduler.js#L483). This 5000ms is _supposed to be the maximum expiration time_, but in this fork it _ends up being the minimum callback time_ (unless another state update forces us to sync-flush pending callbacks). Since `useEffect` is passive, this means React won't call it until after a few seconds have passed.

I assume we (React team) haven't noticed this because we're typically using the `postMessage` implementation, but with JavaScriptCore– React Native ends up using the `setTimeout` fork.

I think this was broken by commit 4d17c3f (PR #13740) which intentionally changed the delay under the (mistaken) assumption that it would only impact unit tests.

If we want to avoid using `setTimeout` for React Native, I think we'll need to add a `scheduler` implementation that does not require `postMessage` or `MessageChannel`, since JavaScript Core does not support either.

**Comment by bvaughn** [comment]
cc @acdlite

**Comment by bvaughn** [comment]
A repro case for this bug can be found [in this Gist](https://gist.github.com/bvaughn/b590a6bfe9ae06fd100dea2a77ab3775). Note that the bug will only repro if you don't connect to a remote debugger (since connecting to Chrome will cause `scheduler` to use the `postMessage` API). This bug can also be reproduced in `react-dom` if you force `scheduler` to use the `setTimeout` API.

**Comment by brunolemos** [comment]
Great that you found the cause!

In addition to this delay, another [reported issue](https://github.com/facebook/react-native/issues/21967#issuecomment-438483674) is that, when js debug is enabled, the useEffect is one render late:

<img height="400" src="https://user-images.githubusercontent.com/619186/48457081-7ba91280-e7a8-11e8-893b-dca30f03e351.gif" />

I didn't wait 5 seconds on this test so it might be related.

**Comment by bvaughn** [comment]
> the useEffect is one render late

I think this is actually the same thing. It would be called after 5 seconds, but you're clicking too soon– in which case, React _synchronously calls it_ before starting on the next render so it seems to be "late" but really, if you were logging the sequence, you'd see that it's called _before_ the follow up render.

**Comment by brunolemos** [comment]
Yeah I think so too

**Comment by brunolemos** [comment]
Now I'm thinking it's not the same thing.

- Like you said, when js debug is enabled it uses the postMessage implementation not the Timeout one.
- I just rewrote the case and after waiting 5 seconds it did not trigger the useEffect

Try the code:

```jsx
import React, { useEffect, useState } from 'react'
import { Button, View } from 'react-native'

export function App() {
  const [count, setCount] = useState(0)

  useEffect(
    () => {
      alert(`Count: ${count}`)
    },
    [count],
  )

  return (
    <View style={{ flex: 1, alignItems: 'center', justifyContent: 'center' }}>
      <Button
        onPress={() => setCount(count + 1)}
        title={`Increment (${count})`}
      />
    </View>
  )
}
```

**Comment by slorber** [comment]
hey, just want to report that the issues I encountered were on Android so both platforms are probably broken

**Comment by bvaughn** [comment]
Interesting. I assume that means that the Android platform is also using the `setTimeout` branch of `scheduler`. If not... 😄

**Comment by bvaughn** [comment]
Okay. I didn't realize but RN is using JSC on Android as well, so it's not a surprise that it impacts both environments. I've updated the description of this issue.

**Comment by jas99** [comment]
@bvaughn Do you think this will be resolved before new year?
 I don't wanna start my new project yuckky.. Classes :laughing: 

Thx

**Comment by bvaughn** [comment]
react 16.7 has a couple of alphas. The first one depends on `scheduler@^0.11.0-alpha` but the others all depend on `scheduler@^0.12.0-alpha`. I'm not sure why this is, but I see that there are some significant changes between the two scheduler alphas– and I feel like back porting my fix to `0.11.0-alpha` would require extra work and testing.

**So for now, I'm only going to patch fix `scheduler@0.12.0-alpha`.**

To use `useEffect` in your React Native app safely, just make sure you're using `react@^16.7.0-alpha.2` and `scheduler@^0.12.0-alpha.3`.

The patch release has now been published.

**Comment by bvaughn** [comment]
I also published the fix to `scheduler@0.11.3` since it potentially impacts the edge case of react-dom `ConcurrentMode` and IE9
